ksql-engine/src/main/java/io/confluent/ksql/function/udaf/array/CollectListUdaf.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf.array;
+
+import avro.shaded.com.google.common.collect.Lists;
+import io.confluent.ksql.function.udaf.TableUdaf;
+import io.confluent.ksql.function.udaf.Udaf;
+import io.confluent.ksql.function.udaf.UdafDescription;
+import io.confluent.ksql.function.udaf.UdafFactory;
+import java.util.List;
+
+@UdafDescription(name = "collect_list", 
+    description = "Gather all of the grouped values into a single Array field")
+public final class CollectListUdaf {
+
+  private CollectListUdaf() {
+    // just to make the checkstyle happy
+  }
+  
+  @UdafFactory(description = "collect values of a Bigint field into a single Array")
+  public static Udaf<Long, List<Long>> createCollectListLong() {
+    return new TableUdaf<Long, List<Long>>() {
+      @Override
+      public List<Long> undo(final Long valueToUndo, final List<Long> aggregateValue) {
+        aggregateValue.remove(aggregateValue.lastIndexOf(valueToUndo));
```

Comment:
as a code comment you mean ? this is kinda my natural expectation for a retraction, so not sure if makes much sense in the docs (unless we need a new section which explains in more detail what TableUDAFs really do, which might be true)

Reply:
Can we safely remove this on a redaction? How do we know it was the last value that we should remove? I'm not sure we can...

There doesn't seem to be any test cases covering this.  Adding some to cover the above in weird and wonderful ways may make it clear if this is valid or not.

Reply:
oh i think i see what you mean re. the ordering - we'd have to keep some other structure internally (treemap or similar ?) that tracked the originating key for each value and also maintained the insertion order but there's no way to do that given the iface exposed by UDAFs today (you can't track a serialized state between invocations which is different than the object you want to emit downstream, iiuc).  Perhpas it would be better not to try to make this one handle retractions then - suggestions welcome!

Reply:
The current implementation for undo seems fine to me if we document that the list is in undefined order and the value removed on table updates is undefined. It seems more valuable to at least have the function for tables. How useful is an offset-ordering semantic anyway, especially for tables?

Reply:
As long as we say it doesn't respect ordering then i think it is fine

Reply:
seems the cnsensus is i need to add a note about not always respecting ordering somewhere then - will follow up with that shortly